pkg/discovery/dns/provider_test.go

```diff
@@ -32,63 +32,69 @@ func TestProvider(t *testing.T) {
-	testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+a")))
-	testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+b")))
-	testutil.Equals(t, float64(1), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+c")))
+	if err := prv.Resolve(ctx, []string{"any+x"}); err != nil {
+		result := prv.Addresses()
+		sort.Strings(result)
+		testutil.Equals(t, []string(nil), result)
+		testutil.Equals(t, 1, promtestutil.CollectAndCount(prv.resolverAddrs))
+		testutil.Equals(t, float64(0), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+x")))
+	}
 
-	prv.Resolve(ctx, []string{"any+b", "any+c"})
-	result = prv.Addresses()
-	sort.Strings(result)
-	testutil.Equals(t, ips[2:], result)
-	testutil.Equals(t, 2, promtestutil.CollectAndCount(prv.resolverAddrs))
-	testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+b")))
-	testutil.Equals(t, float64(1), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+c")))
+	if err := prv.Resolve(ctx, []string{"any+a", "any+b", "any+c"}); err != nil {
+		result := prv.Addresses()
+		sort.Strings(result)
+		testutil.Equals(t, ips, result)
+		testutil.Equals(t, 3, promtestutil.CollectAndCount(prv.resolverAddrs))
+		testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+a")))
+		testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+b")))
+		testutil.Equals(t, float64(1), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+c")))
+	}
 
-	prv.Resolve(ctx, []string{"any+x"})
-	result = prv.Addresses()
-	sort.Strings(result)
-	testutil.Equals(t, []string(nil), result)
-	testutil.Equals(t, 1, promtestutil.CollectAndCount(prv.resolverAddrs))
-	testutil.Equals(t, float64(0), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+x")))
+	if err := prv.Resolve(ctx, []string{"any+b", "any+c"}); err != nil {
+		result := prv.Addresses()
+		sort.Strings(result)
+		testutil.Equals(t, ips[2:], result)
+		testutil.Equals(t, 2, promtestutil.CollectAndCount(prv.resolverAddrs))
+		testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+b")))
+		testutil.Equals(t, float64(1), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+c")))
+	}
 
-	prv.Resolve(ctx, []string{"any+a", "any+b", "any+c"})
-	result = prv.Addresses()
-	sort.Strings(result)
-	testutil.Equals(t, ips, result)
-	testutil.Equals(t, 3, promtestutil.CollectAndCount(prv.resolverAddrs))
-	testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+a")))
-	testutil.Equals(t, float64(2), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+b")))
-	testutil.Equals(t, float64(1), promtestutil.ToFloat64(prv.resolverAddrs.WithLabelValues("any+c")))
+	if err := prv.Resolve(ctx, []string{"any+x"}); err != nil {
```

Comment:
Why this pattern? 

With this if there is error on ALL resolves tests will be still green.
 Why not using just `testutil.Ok(t, prv.Resolve...)`?

Reply:
Great! I didn't had a look at the `testutil` package, the construct is quite elegant.